docs/using-connect/integration.mdx

```diff
@@ -0,0 +1,374 @@
+	}
+
+	// If app level instrumentation is enabled, then wrap the oracle service with a metrics client
+	// to get metrics on the oracle service (for ABCI++). This will allow the instrumentation to track
+	// latency in VerifyVoteExtension requests and more.
+	oracleMetrics, err := servicemetrics.NewMetricsFromConfig(cfg, app.ChainID())
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Create the oracle service.
+	oracleClient, err := oracleclient.NewClientFromConfig(
+		cfg,
+		app.Logger().With("client", "oracle"),
+		oracleMetrics,
+	)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Connect to the oracle service (default timeout of 5 seconds).
+	go func() {
+		if err := oracleClient.Start(context.Background()); err != nil {
+			app.Logger().Error("failed to start oracle client", "err", err)
+			panic(err)
+		}
+
+		app.Logger().Info("started oracle client", "address", cfg.OracleAddress)
+	}()
+	return oracleClient, oracleMetrics, nil
+}
+```
+## ABCI and Vote Extensions
+Configure the ABCI methods and vote extensions.
+
+Define a method to contain the logic where these will be configured.
+```go oracle.go
+func (app *App) initializeABCIExtensions(oracleClient oracleclient.OracleClient, oracleMetrics servicemetrics.Metrics) {}
+```
+
+Within this method, do the following:
+- **Setup Proposal Handler:** This handler will be used in `PrepareProposal` and `ProcessProposal` to fill proposals with the oracle data.
+- **Set PreBlocker:** The Application's `PreBlocker` will be configured to write price data to state before transactions are executed.
+- **Set Vote Extensions:** Set the vote extension handlers on the Application that will handle adding price data to the node's consensus votes.
+
+Start with setting up the proposal handler.
+
+```go oracle.go
+package app
+
+import (
+	oracleclient "github.com/skip-mev/slinky/service/clients/oracle"
+    servicemetrics "github.com/skip-mev/slinky/service/metrics"
+
+    "github.com/cosmos/cosmos-sdk/baseapp"
+    "github.com/skip-mev/slinky/abci/proposals"
+    compression "github.com/skip-mev/slinky/abci/strategies/codec"
+    "github.com/skip-mev/slinky/abci/strategies/currencypair"
+    "github.com/skip-mev/slinky/abci/ve"
+)
+
+func (app *App) initializeABCIExtensions(oracleClient oracleclient.OracleClient, oracleMetrics servicemetrics.Metrics) {
+    // Create the proposal handler that will be used to fill proposals with
+	// transactions and oracle data.
+	proposalHandler := proposals.NewProposalHandler(
+		app.Logger(),
+		baseapp.NoOpPrepareProposal(),
+		baseapp.NoOpProcessProposal(),
+		ve.NewDefaultValidateVoteExtensionsFn(app.StakingKeeper),
+		compression.NewCompressionVoteExtensionCodec(
+			compression.NewDefaultVoteExtensionCodec(),
+			compression.NewZLibCompressor(),
+		),
+		compression.NewCompressionExtendedCommitCodec(
+			compression.NewDefaultExtendedCommitCodec(),
+			compression.NewZStdCompressor(),
+		),
+		currencypair.NewDeltaCurrencyPairStrategy(app.OracleKeeper),
```

Comment:
It is critical to note that the strategy cannot be shared between the vote extension and proposal handlers. They have to be instantiated separately otherwise the caching system gets all weird.

Reply:
that seems really easy to shoot yourself in the foot lol, i'll open a ticket to look into that

Reply:
https://linear.app/skip/issue/BLO-1578/currency-pair-strategy-objects-cannot-be-reused